### ISSUE TITLE
Memory caching calls slow for NumPy arguments

### ISSUE THREAD

**Issue by AdamGleave** [joblib/joblib#683]
When using Memory.cache, Joblib writes out a metadata file for each cached function call including its input argument. While Joblib uses an efficient NumPy pickler for *output* arrays, it calls repr(x) on each input argument x on arrays appearing in the *input* arguments. This is exceedingly slow, as the below benchmark demonstrates.

```python3
from contextlib import contextmanager                                                                                                                                                                 
from joblib import Memory
import numpy as np
import time

memory = Memory(cachedir="tmp_cache")
A = [np.random.randn(100) for _ in range(1000)]

@memory.cache
def f(x):
    return A

@memory.cache
def g(M):
    return M

@contextmanager
def timer(label):
    start = time.time()
    yield
    end = time.time()
    print('{} - {}s'.format(label, end - start))

with timer("f miss"):
    X = f(42)
with timer("f cached"):
    X = f(42)
with timer("g miss"):
    Y = g(X)
with timer("g cached"):
    Y = g(X)
```

Running this with joblib 0.11 on Python 3.6.5 on a Linux machine gives:

```
f miss - 0.07152724266052246s
f cached - 0.05348658561706543s
g miss - 1.5449304580688477s
g cached - 0.14382505416870117s    
```

Note that although ```f``` and ```g``` return the same value, calling ```g``` takes an order of magnitude more time when there is a cache miss (writing the metadata file). Subsequent (cached) calls take a similar amount of time. I also observe the following warning: 

```
joblib_benchmark.py:29: UserWarning: Persisting input arguments took 1.10s to run.
If this happens often in your code, it can cause performance problems 
(results will be correct in all cases). 
The reason for this is probably some large input arguments for a wrapped
 function (e.g. large strings).
THIS IS A JOBLIB ISSUE. If you can, kindly provide the joblib's team with an
 example so that they can fix the problem.
```

As a workaround I am currently using ```np.set_printoptions``` to reduce the verbosity of repr(x) on a NumPy array x. We could potentially use the same trick in _persist_input (restoring the original options with a {get,set}_printoptions pair), but this does not feel very satisfactory. It's unclear to me what the design criteria for metadata.json is -- we could pickle the arguments efficiently, but perhaps it's desirable they be human readable? I'd be happy to have a go at fixing this with some guidance on the design space.

**Comment by tritemio** [comment]
I'm also getting this warning running a cached function that takes a large pandas Series as input. Reducing the number of rows printed by default by pandas helped.

**Comment by neergaard** [comment]
@AdamGleave Where did you specify the print options? I am running into the same problem, but your solution did not work for me. I could just be implementing it wrong.

**Comment by AdamGleave** [comment]
> edit: I tried two solutions 1) to call `np.set_printoptions` in the file that contains the called function, and 2) decorating the call to Parallel which also calls the cached function with `with np.printoptions`, but neither did work.

What you're doing sounds like it should work. I'd suggest trying some micro-benchmarks of `repr(x)` where `x` is something representative of your input, and change the settings until you get reasonable performance on that.

**Comment by anntzer** [comment]
Here's another self-contained repro to trigger the warning:
```python
import numpy as np
from joblib import Memory

mem = Memory("/tmp/joblib-cache")

@mem.cache
def func(x): return np.mean(x)

np.random.randn(int(1e8)).tofile("/tmp/rand.bin")  # adjust size accordingly
x = np.memmap("/tmp/rand.bin")
func(x)
```
In this case, the problem arises from the fact that computing the repr of the memmap is slow.

**Comment by tomMoral** [comment]
I agree that the warning is unclear and we could actually update it.

The purpose of the warning is to help people detect that the caching is not free. It comes with the overhead of hashing all input arguments and checking if they are in the database.
To make the cache work, we actually need to check all the input arguments (that means all entries in a numpy array) because if only one element is different, you don't want to get the same (wrong) result. That means that the hashing is linear in the size of the input and I don't see clear ways to make it faster.

If the function is faster that the overhead, the cache should not be used, or used differently.
So we should update the warning to point to two mitigations:

- If the function call is smaller than the overhead, one should not use the memory cache at all, as we cannot do better for this (the hashing process is linear with the size of the input so we cannot go faster).
- But one could reduce this overhead by only caching functions with simple arguments (string, int, ... *i.e.* parameters and not data). Usually, when using the cache, I try to put the cache at a level before I load the data as hashing the data will always be slow.

Not sure if this answer the question or if you see a better way to improve this?

**Comment by tomMoral** [comment]
Actually there is two part here:
- The time taken by the serialisation (in call `_get_output_identifiers`). If this is slower than the function call, it makes no sense to use the cache. However, this part is also call when the cache is hit, and in the OP case, this is not a big problem (0.15s) compared to the first call that is very slow.
- Then the `store_metadata` stores a summary of the argument used to call the function `input_repr`. This part, which is the one only called when the cache is missed is the one responsible for having a very large timing (larger than the hash) which explains why the OP observe a `1.5s` overhead that disappear when he change the cost of representing the numpy array.

As this metadata is not used a lot, maybe we could compute a custom representation for the large arguments that would be cheaper here to mitigate the apparition of the warning.

also, I think having different warning for when the hashing is too slow or when the metadata storage is too slow would help understand which case is really problematic and better indicate the problem to the user. If we have slow hashing, there is not much we can do. If it is on the metadata, then maybe we should reduce the metada as currently, they are not used anywhere I guess?

**Comment by alexisthual** [comment]
+1 for two different warnings!

Also, this warning popped up in one of my scripts where I cache a function which loads a gifti map located at a specific path.
It feels weird that hashing a file path would take more time than loading the whole file. Am I missing something?
